c.alibaba.druid.filter.stat.StatFilter   : merge sql error, dbType highgo

2023-05-31 10:00:23.089 ERROR 3432 --- [4-1685498417769] c.alibaba.druid.filter.stat.StatFilter   : merge sql error, dbType highgo, druid-1.2.16, sql : SELECT replace(id,'-','') id,
            ? `type`,
            catalog_id,
            catalog_name,
            items,
            supply_dept_name,
            apply_number,
            catalog_update_time
        FROM
            (
            SELECT

                 
                        cast(gen_random_uuid() as text) id,
                     
                a.id catalog_id,
                a.catalog_name,
                GROUP_CONCAT( b.item_name SEPARATOR '、' ) items,
                a.dept_name supply_dept_name,
                case when a.apply_number is null then 0 else a.apply_number end apply_number,
                a.update_time catalog_update_time
            FROM
                tsds_resource_catalog a
                LEFT JOIN tsds_resource_catalog_item b ON a.id = b.catalog_id
                AND b.is_deleted = '0'
                where a.is_deleted='0' and a.is_offline='0'
            GROUP BY
                a.id
            ) t
        ORDER BY
            apply_number DESC
            LIMIT 5

com.alibaba.druid.sql.parser.ParserException: syntax error, expect ), actual IDENTIFIER pos 472, line 18, column 44, token IDENTIFIER SEPARATOR
	at com.alibaba.druid.sql.parser.SQLExprParser.accept(SQLExprParser.java:4610)
	at com.alibaba.druid.sql.parser.SQLExprParser.methodRest(SQLExprParser.java:1925)
	at com.alibaba.druid.sql.parser.SQLExprParser.parseSelectItem(SQLExprParser.java:6290)
	at com.alibaba.druid.sql.parser.SQLSelectParser.parseSelectList(SQLSelectParser.java:1068)
	at com.alibaba.druid.sql.parser.SQLSelectParser.query(SQLSelectParser.java:462)
	at com.alibaba.druid.sql.parser.SQLSelectParser.select(SQLSelectParser.java:61)
	at com.alibaba.druid.sql.parser.SQLSelectParser.parseTableSource(SQLSelectParser.java:1105)
	at com.alibaba.druid.sql.parser.SQLSelectParser.parseFrom(SQLSelectParser.java:1096)
	at com.alibaba.druid.sql.parser.SQLSelectParser.query(SQLSelectParser.java:473)
	at com.alibaba.druid.sql.parser.SQLSelectParser.select(SQLSelectParser.java:61)
	at com.alibaba.druid.sql.parser.SQLStatementParser.parseSelect(SQLStatementParser.java:4380)
	at com.alibaba.druid.sql.parser.SQLStatementParser.parseStatementList(SQLStatementParser.java:200)
	at com.alibaba.druid.sql.parser.SQLStatementParser.parseStatementList(SQLStatementParser.java:101)
	at com.alibaba.druid.sql.visitor.ParameterizedOutputVisitorUtils.parameterize(ParameterizedOutputVisitorUtils.java:163)
	at com.alibaba.druid.sql.visitor.ParameterizedOutputVisitorUtils.parameterize(ParameterizedOutputVisitorUtils.java:134)
	at com.alibaba.druid.filter.stat.StatFilter.mergeSql(StatFilter.java:148)
	at com.alibaba.druid.filter.stat.StatFilter.createSqlStat(StatFilter.java:672)
	at com.alibaba.druid.filter.stat.StatFilter.statementPrepareAfter(StatFilter.java:325)
	at com.alibaba.druid.filter.FilterEventAdapter.connection_prepareStatement(FilterEventAdapter.java:118)
	at com.alibaba.druid.filter.FilterChainImpl.connection_prepareStatement(FilterChainImpl.java:531)
	at com.alibaba.druid.proxy.jdbc.ConnectionProxyImpl.prepareStatement(ConnectionProxyImpl.java:326)
	at com.alibaba.druid.pool.DruidPooledConnection.prepareStatement(DruidPooledConnection.java:362)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.instantiateStatement(PreparedStatementHandler.java:86)
	at org.apache.ibatis.executor.statement.BaseStatementHandler.prepare(BaseStatementHandler.java:88)
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.prepare(RoutingStatementHandler.java:59)
	at org.apache.ibatis.executor.SimpleExecutor.prepareStatement(SimpleExecutor.java:87)
	at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:62)
	at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:325)
	at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:156)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:109)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:89)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:151)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:145)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:140)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:427)
	at com.sun.proxy.$Proxy92.selectList(Unknown Source)
	at org.mybatis.spring.SqlSessionTemplate.selectList(SqlSessionTemplate.java:224)
	at org.apache.ibatis.binding.MapperMethod.executeForMany(MapperMethod.java:147)
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:80)
	at org.apache.ibatis.binding.MapperProxy$PlainMethodInvoker.invoke(MapperProxy.java:145)
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:86)
	at com.sun.proxy.$Proxy178.list(Unknown Source)
	at com.xxl.job.executor.biz.service.impl.PortalServiceImpl.updatePortalCatalog(PortalServiceImpl.java:65)
	at com.xxl.job.executor.biz.service.impl.PortalServiceImpl$$FastClassBySpringCGLIB$$5da5636a.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at com.alibaba.druid.support.spring.stat.DruidStatInterceptor.invoke(DruidStatInterceptor.java:70)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at com.xxl.job.executor.biz.service.impl.PortalServiceImpl$$EnhancerBySpringCGLIB$$9bfe1a2e.updatePortalCatalog(<generated>)
	at com.xxl.job.executor.job.PortalExecutorXxlJob.updatePortalCatalog(PortalExecutorXxlJob.java:31)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.xxl.job.core.handler.impl.MethodJobHandler.execute(MethodJobHandler.java:31)
	at com.xxl.job.core.thread.JobThread.run(JobThread.java:166)

遇到了同样的问题，也是瀚高数据库
`
select o.sku_id skuId,o.`sku_name` skuName,o.order_type orderType,o.sort, group_concat(si.supplier_name order by si.sort separator '、') as supplierName 
from orders as o  
left join sku_items as si on o.id = si.contract_id  
group by o.sku_id,o.`sku_name`,o.order_type,o.sort  
order by o.sort
`
`
Caused by: com.alibaba.druid.sql.parser.ParserException: syntax error, expect ), actual IDENTIFIER pos 132, line 1, column 132, token IDENTIFIER separator
	at com.alibaba.druid.sql.parser.SQLExprParser.accept(SQLExprParser.java:4616)
	at com.alibaba.druid.sql.parser.SQLExprParser.methodRest(SQLExprParser.java:1931)
	at com.alibaba.druid.sql.parser.SQLExprParser.parseSelectItem(SQLExprParser.java:6296)
	at com.alibaba.druid.sql.parser.SQLSelectParser.parseSelectList(SQLSelectParser.java:1068)
	at com.alibaba.druid.sql.dialect.postgresql.parser.PGSelectParser.query(PGSelectParser.java:103)
	at com.alibaba.druid.sql.parser.SQLSelectParser.select(SQLSelectParser.java:61)
	at com.alibaba.druid.sql.dialect.postgresql.parser.PGSQLStatementParser.parseSelect(PGSQLStatementParser.java:416)
	at com.alibaba.druid.sql.dialect.postgresql.parser.PGSQLStatementParser.parseSelect(PGSQLStatementParser.java:30)
	at com.alibaba.druid.sql.parser.SQLStatementParser.parseStatementList(SQLStatementParser.java:200)
	at com.alibaba.druid.sql.parser.SQLStatementParser.parseStatementList(SQLStatementParser.java:112)
	at com.alibaba.druid.wall.WallProvider.checkInternal(WallProvider.java:618)
	at com.alibaba.druid.wall.WallProvider.check(WallProvider.java:572)
	at com.alibaba.druid.wall.WallFilter.checkInternal(WallFilter.java:806)
	... 225 more
`

--- a/core/src/main/java/com/alibaba/druid/sql/ast/expr/SQLHexExpr.java
+++ b/core/src/main/java/com/alibaba/druid/sql/ast/expr/SQLHexExpr.java
@@ -33,8 +33,13 @@ public class SQLHexExpr extends SQLExprImpl implements SQLLiteralExpr, SQLValuab
     }
 
     public void output(StringBuilder buf) {
-        buf.append("0x");
-        buf.append(this.hex);
+        if (this.hex == null || this.hex.isEmpty()) {
+            // 0x must have at least one digit; an empty hex literal is written X'' (issue #5244)
+            buf.append("X''");
+        } else {
+            buf.append("0x");
+            buf.append(this.hex);
+        }
 
         String charset = (String) getAttribute("USING");
         if (charset != null) {

--- a/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/ast/stmt/PGSelectQueryBlock.java
+++ b/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/ast/stmt/PGSelectQueryBlock.java
@@ -136,7 +136,7 @@ public class PGSelectQueryBlock extends SQLSelectQueryBlock implements PGSQLObje
 
     public static class ForClause extends PGSQLObjectImpl {
         public static enum Option {
-            UPDATE, SHARE
+            UPDATE, SHARE, KEY_SHARE, NO_KEY_UPDATE
         }
 
         private List<SQLExpr> of = new ArrayList<SQLExpr>(2);

--- a/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/parser/PGSelectParser.java
+++ b/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/parser/PGSelectParser.java
@@ -209,8 +209,19 @@ public class PGSelectParser extends SQLSelectParser {
             } else if (lexer.token() == Token.SHARE) {
                 forClause.setOption(PGSelectQueryBlock.ForClause.Option.SHARE);
                 lexer.nextToken();
+            } else if (lexer.token() == Token.KEY) {
+                // FOR KEY SHARE
+                lexer.nextToken();
+                accept(Token.SHARE);
+                forClause.setOption(PGSelectQueryBlock.ForClause.Option.KEY_SHARE);
+            } else if (lexer.identifierEquals("NO")) {
+                // FOR NO KEY UPDATE
+                lexer.nextToken();
+                accept(Token.KEY);
+                accept(Token.UPDATE);
+                forClause.setOption(PGSelectQueryBlock.ForClause.Option.NO_KEY_UPDATE);
             } else {
-                throw new ParserException("expect 'FIRST' or 'NEXT'. " + lexer.info());
+                throw new ParserException("expect UPDATE, SHARE, KEY SHARE or NO KEY UPDATE. " + lexer.info());
             }
 
             if (lexer.token() == Token.OF) {

--- a/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/parser/PGSelectParser.java
+++ b/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/parser/PGSelectParser.java
@@ -225,6 +225,7 @@ public class PGSelectParser extends SQLSelectParser {
             }
 
             if (lexer.token() == Token.OF) {
+                lexer.nextToken();
                 for (; ; ) {
                     SQLExpr expr = this.createExprParser().expr();
                     forClause.getOf().add(expr);

--- a/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/visitor/PGOutputVisitor.java
+++ b/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/visitor/PGOutputVisitor.java
@@ -77,6 +77,10 @@ public class PGOutputVisitor extends SQLASTOutputVisitor implements PGASTVisitor
             print0(ucase ? "UPDATE" : "update");
         } else if (ForClause.Option.SHARE.equals(x.getOption())) {
             print0(ucase ? "SHARE" : "share");
+        } else if (ForClause.Option.KEY_SHARE.equals(x.getOption())) {
+            print0(ucase ? "KEY SHARE" : "key share");
+        } else if (ForClause.Option.NO_KEY_UPDATE.equals(x.getOption())) {
+            print0(ucase ? "NO KEY UPDATE" : "no key update");
         }
 
         if (x.getOf().size() > 0) {

--- a/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/visitor/PGOutputVisitor.java
+++ b/core/src/main/java/com/alibaba/druid/sql/dialect/postgresql/visitor/PGOutputVisitor.java
@@ -84,10 +84,10 @@ public class PGOutputVisitor extends SQLASTOutputVisitor implements PGASTVisitor
         }
 
         if (x.getOf().size() > 0) {
-            print(' ');
+            print0(ucase ? " OF " : " of ");
             for (int i = 0; i < x.getOf().size(); ++i) {
                 if (i != 0) {
-                    println(", ");
+                    print0(", ");
                 }
                 x.getOf().get(i).accept(this);
             }

--- a/core/src/main/java/com/alibaba/druid/sql/visitor/SQLASTOutputVisitor.java
+++ b/core/src/main/java/com/alibaba/druid/sql/visitor/SQLASTOutputVisitor.java
@@ -4429,8 +4429,14 @@ public class SQLASTOutputVisitor extends SQLASTVisitorAdapter implements Paramet
             return false;
         }
 
-        print0("0x");
-        print0(x.getHex());
+        String hex = x.getHex();
+        if (hex == null || hex.isEmpty()) {
+            // 0x must have at least one digit; an empty hex literal is written X'' (issue #5244)
+            print0("X''");
+        } else {
+            print0("0x");
+            print0(hex);
+        }
 
         String charset = (String) x.getAttribute("USING");
         if (charset != null) {

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/clickhouse/issues/Issue5235.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/clickhouse/issues/Issue5235.java
@@ -1,0 +1,28 @@
+package com.alibaba.druid.bvt.sql.clickhouse.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * ClickHouse array-predicate functions hasAll / hasAny with array literal arguments.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5235">Issue #5235</a>
+ */
+public class Issue5235 {
+    @Test
+    public void test_has_all_has_any() {
+        for (String fn : new String[]{"hasAll", "hasAny"}) {
+            String sql = "SELECT " + fn + "([1,2,3], [1,2]) FROM t";
+            List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.clickhouse);
+            assertEquals(1, stmts.size());
+            assertTrue(SQLUtils.toSQLString(stmts.get(0), DbType.clickhouse).contains(fn + "("));
+        }
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/mysql/issues/Issue5244.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/mysql/issues/Issue5244.java
@@ -1,0 +1,37 @@
+package com.alibaba.druid.bvt.sql.mysql.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * MySQL empty hex literal X'' must not be rendered as the invalid 0x (0x needs at least one digit).
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5244">Issue #5244</a>
+ */
+public class Issue5244 {
+    private static String rt(String sql) {
+        List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.mysql);
+        assertEquals(1, stmts.size());
+        return SQLUtils.toSQLString(stmts.get(0), DbType.mysql).replaceAll("\\s+", " ").trim();
+    }
+
+    @Test
+    public void test_empty_hex_literal() {
+        String out = rt("INSERT INTO `t` (`id`, `v`) VALUES (NULL, X'')");
+        assertTrue(out.contains("X''"), out);
+        assertFalse(out.contains("0x)") || out.contains("0x "), "must not emit a bare 0x: " + out);
+    }
+
+    @Test
+    public void test_non_empty_hex_literal_unchanged() {
+        assertTrue(rt("INSERT INTO t (v) VALUES (X'41')").contains("0x41"));
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/mysql/issues/Issue5244.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/mysql/issues/Issue5244.java
@@ -3,6 +3,7 @@ package com.alibaba.druid.bvt.sql.mysql.issues;
 import com.alibaba.druid.DbType;
 import com.alibaba.druid.sql.SQLUtils;
 import com.alibaba.druid.sql.ast.SQLStatement;
+import com.alibaba.druid.sql.ast.expr.SQLHexExpr;
 import org.junit.jupiter.api.Test;
 
 import java.util.List;
@@ -33,5 +34,13 @@ public class Issue5244 {
     @Test
     public void test_non_empty_hex_literal_unchanged() {
         assertTrue(rt("INSERT INTO t (v) VALUES (X'41')").contains("0x41"));
+    }
+
+    @Test
+    public void test_empty_hex_tostring_path() {
+        // SQLObjectImpl.toString() goes through SQLHexExpr.output(), not the visitor
+        assertEquals("X''", new SQLHexExpr("").toString());
+        assertEquals("X''", new SQLHexExpr((String) null).toString());
+        assertEquals("0x41", new SQLHexExpr("41").toString());
     }
 }

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/mysql/issues/Issue5304.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/mysql/issues/Issue5304.java
@@ -1,0 +1,26 @@
+package com.alibaba.druid.bvt.sql.mysql.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * GROUP_CONCAT(expr SEPARATOR '...') with a multi-byte separator.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5304">Issue #5304</a>
+ */
+public class Issue5304 {
+    @Test
+    public void test_group_concat_separator() {
+        String sql = "SELECT GROUP_CONCAT(name SEPARATOR '、') FROM t";
+        List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.mysql);
+        assertEquals(1, stmts.size());
+        assertTrue(SQLUtils.toSQLString(stmts.get(0), DbType.mysql).contains("SEPARATOR '、'"));
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue5203.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue5203.java
@@ -38,4 +38,22 @@ public class Issue5203 {
         assertEquals("SELECT id FROM t FOR UPDATE", rt("SELECT id FROM t FOR UPDATE"));
         assertEquals("SELECT id FROM t FOR SHARE", rt("SELECT id FROM t FOR SHARE"));
     }
+
+    @Test
+    public void test_for_key_share_with_of() {
+        assertEquals("SELECT * FROM t FOR KEY SHARE OF t",
+                rt("SELECT * FROM t FOR KEY SHARE OF t"));
+    }
+
+    @Test
+    public void test_for_no_key_update_with_nowait() {
+        assertEquals("SELECT * FROM t FOR NO KEY UPDATE NOWAIT",
+                rt("SELECT * FROM t FOR NO KEY UPDATE NOWAIT"));
+    }
+
+    @Test
+    public void test_for_key_share_skip_locked() {
+        assertEquals("SELECT * FROM t FOR KEY SHARE SKIP LOCKED",
+                rt("SELECT * FROM t FOR KEY SHARE SKIP LOCKED"));
+    }
 }

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue5203.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue5203.java
@@ -1,0 +1,41 @@
+package com.alibaba.druid.bvt.sql.postgresql.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * PostgreSQL row-locking clauses: FOR UPDATE / FOR NO KEY UPDATE / FOR SHARE / FOR KEY SHARE.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5203">Issue #5203</a>
+ */
+public class Issue5203 {
+    private static String rt(String sql) {
+        List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.postgresql);
+        assertEquals(1, stmts.size());
+        return SQLUtils.toSQLString(stmts.get(0), DbType.postgresql).replaceAll("\\s+", " ").trim();
+    }
+
+    @Test
+    public void test_for_key_share() {
+        assertEquals("SELECT id FROM t WHERE id = 1 FOR KEY SHARE",
+                rt("SELECT id FROM t WHERE id = 1 FOR KEY SHARE"));
+    }
+
+    @Test
+    public void test_for_no_key_update() {
+        assertEquals("SELECT id FROM t WHERE id = 1 FOR NO KEY UPDATE",
+                rt("SELECT id FROM t WHERE id = 1 FOR NO KEY UPDATE"));
+    }
+
+    @Test
+    public void test_for_update_and_share_unchanged() {
+        assertEquals("SELECT id FROM t FOR UPDATE", rt("SELECT id FROM t FOR UPDATE"));
+        assertEquals("SELECT id FROM t FOR SHARE", rt("SELECT id FROM t FOR SHARE"));
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue5266.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue5266.java
@@ -1,0 +1,26 @@
+package com.alibaba.druid.bvt.sql.postgresql.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * PostgreSQL CREATE TABLE with an auto-increment column GENERATED ALWAYS AS IDENTITY.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5266">Issue #5266</a>
+ */
+public class Issue5266 {
+    @Test
+    public void test_generated_always_as_identity() {
+        String sql = "CREATE TABLE t (id integer NOT NULL GENERATED ALWAYS AS IDENTITY, name varchar(10))";
+        List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.postgresql);
+        assertEquals(1, stmts.size());
+        assertTrue(SQLUtils.toSQLString(stmts.get(0), DbType.postgresql).contains("GENERATED ALWAYS AS IDENTITY"));
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue6633.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/postgresql/issues/Issue6633.java
@@ -1,0 +1,56 @@
+package com.alibaba.druid.bvt.sql.postgresql.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import com.alibaba.druid.sql.repository.SchemaRepository;
+import com.alibaba.druid.wall.WallCheckResult;
+import com.alibaba.druid.wall.spi.PGWallProvider;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+/**
+ * Visiting a PostgreSQL-specific expression (EXTRACT, ::cast, ...) with a non-PG visitor used to
+ * throw ClassCastException (SQLEvalVisitorImpl / SchemaResolveVisitor cannot be cast to PGASTVisitor),
+ * e.g. during WallFilter eval or schema resolve. PGExprImpl now checks instanceof before casting.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/6633">Issue #6633</a>
+ * @see <a href="https://github.com/alibaba/druid/issues/6158">Issue #6158</a>
+ * @see <a href="https://github.com/alibaba/druid/issues/5201">Issue #5201</a>
+ */
+public class Issue6633 {
+    private static final String[] SQLS = {
+            "select * from a where extract(epoch from current_date)::bigint > 1",
+            "select EXTRACT(EPOCH FROM COALESCE(a, now())) - EXTRACT(EPOCH FROM b) as s from t",
+    };
+
+    @Test
+    public void test_resolve_does_not_throw_cce() {
+        for (String sql : SQLS) {
+            SchemaRepository repo = new SchemaRepository(DbType.postgresql);
+            List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.postgresql);
+            for (SQLStatement stmt : stmts) {
+                repo.resolve(stmt);
+            }
+        }
+    }
+
+    @Test
+    public void test_wall_eval_does_not_throw_cce() {
+        PGWallProvider provider = new PGWallProvider();
+        for (String sql : SQLS) {
+            WallCheckResult result = provider.check(sql);
+            assertNotNull(result);
+        }
+    }
+
+    @Test
+    public void test_select_after_extract_cast_parses() {
+        List<SQLStatement> stmts = SQLUtils.parseStatements(SQLS[0], DbType.postgresql);
+        assertEquals(1, stmts.size());
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/spark/issues/Issue5269.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/spark/issues/Issue5269.java
@@ -1,0 +1,26 @@
+package com.alibaba.druid.bvt.sql.spark.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Spark SQL file-format table reference, e.g. FROM parquet.`oss://bucket/path`.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5269">Issue #5269</a>
+ */
+public class Issue5269 {
+    @Test
+    public void test_format_path_table_source() {
+        String sql = "select sum(show_cnt) from parquet.`oss://bucket/path` limit 1000";
+        List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.spark);
+        assertEquals(1, stmts.size());
+        assertTrue(SQLUtils.toSQLString(stmts.get(0), DbType.spark).contains("parquet.`oss://bucket/path`"));
+    }
+}

--- a/core/src/test/java/com/alibaba/druid/bvt/sql/sqlserver/issues/Issue5211.java
+++ b/core/src/test/java/com/alibaba/druid/bvt/sql/sqlserver/issues/Issue5211.java
@@ -1,0 +1,36 @@
+package com.alibaba.druid.bvt.sql.sqlserver.issues;
+
+import com.alibaba.druid.DbType;
+import com.alibaba.druid.sql.SQLUtils;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * SQL Server OUTER APPLY / CROSS APPLY.
+ *
+ * @see <a href="https://github.com/alibaba/druid/issues/5211">Issue #5211</a>
+ */
+public class Issue5211 {
+    private static String rt(String sql) {
+        List<SQLStatement> stmts = SQLUtils.parseStatements(sql, DbType.sqlserver);
+        assertEquals(1, stmts.size());
+        return SQLUtils.toSQLString(stmts.get(0), DbType.sqlserver).replaceAll("\\s+", " ").trim();
+    }
+
+    @Test
+    public void test_outer_apply() {
+        assertTrue(rt("select t1.name from tableA as t1 outer apply (select * from tableB t2 where t2.id = 1) where t1.name = ?")
+                .contains("OUTER APPLY"));
+    }
+
+    @Test
+    public void test_cross_apply() {
+        assertTrue(rt("select t1.name from tableA t1 cross apply (select * from tableB t2 where t2.id = 1) t")
+                .contains("CROSS APPLY"));
+    }
+}

